config/user_agent_filter.go

```diff
@@ -0,0 +1,35 @@
+package config
+
+import (
+	"golang.ngrok.com/ngrok/internal/pb"
+)
+
+// UserAgentFilter is a pair of strings slices that allow/deny traffic to an endpoint
+type userAgentFilter struct {
+	// slice of regex strings for allowed user agents
+	Allow []string
+	// slice of regex strings for denied user agents
+	Deny []string
+}
+
+func (b *userAgentFilter) toProtoConfig() *pb.MiddlewareConfiguration_UserAgentFilter {
+	if b == nil {
+		return nil
+	}
+	return &pb.MiddlewareConfiguration_UserAgentFilter{
+		Allow: b.Allow,
+		Deny:  b.Deny,
+	}
+}
+
+// WithUserAgentFilter configures user-agent filter from passed parameters.
+func WithUserAgentFilter(allow []string, deny []string) HTTPEndpointOption {
```

Comment:
To repeat the same comment as the last two reviews, we should document what the arguments to this function does as part of the godoc comment.

Right now, there's nowhere in godoc that makes it clear to the caller that these are regular expressions, nor what the semantics are.

I'll also point out that we could take a `[]*regexp.Regexp` type if we wanted, for type safety (since `.String()` can recover the regexp), but we don't have to do that.

We do have to get the godoc comment right / usable, or at least link to upstream docs.

Reply:
mb lol